man/man1/snapmnmx.1
.TH SNAPMNMX 1NEMO "17 August 2022"

.SH "NAME"
snapmnmx \- show statistics of snapshot variables, optionally output a single min or max particle

.SH "SYNOPSIS"
.PP
\fBsnapmnmx in=\fPsnap_in  [parameter=value]

.SH "DESCRIPTION"
\fIsnapmnmx\fP produces a table of a number of selected statistics
of requested quantities of the bodies in a snapshot. It can also add
the selected times to the table.
.PP
A special case where \fBout=\fP is given, and
either just the \fBmin\fP or \fBmax\fP was selected
of the variable will cause this particle to be output in a single particle snapshot.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
Input file, in \fIsnapshot(5NEMO)\fP format [no default].
.TP
\fBvar=\fIvar_list\fP
List of data-items for which statistics will be output;
any of the \fIbodytrans(3NEM0)\fP
functions can be used in an arbitry expression.
The order in the table is the order of the variables given 
within this keyword, within each selected mode (see next keyword).
[default: \fBx,y,z\fP].
.TP
\fBmode=\fImode_list\fP
List modes of output, which can be combined. Valid 
options are: \fBtime, min, max, mean\fP and \fBsigma\fP. 
Multiple options must be separated by a comma. The columns
are printed in the order of the list above. However,
the \fBvar=\fP is the ``fastest'' variable of this two dimensional
matrix. Example: with the defaults (\fBmode=min,max\fP 
and \fBvar=x,y,z\fP)
the program would print \fIxmin, ymin, zmin, xmax, ymax, xmax\fP.
[default: \fBmin,max\fP].
.TP
\fBtimes=\fItimes-string\fP
Time values/intervals of which snapshots should be used. 
[default: \fBall\fP]
.TP
\fBformat=\fIformat=-string\fP
Format string (in the C language notation) of the output numbers.
If no space is used in the format description, the program will
add one for you. [default: \fB%g\fP (free format)].
.TP
\fBout=\fP
If give, this will be the output file of the particle for which the min or max was selected.
Only a single particle can be output here.

.SH "EXAMPLES"
.nf
    # make a Plummer sphere with 100 particles and given seed
    mkplummer p100 100 seed=123
    
    #  find the min x,y,z and max x,y,z; the bounding box of this model
    snapmnmx p100
    -4.67407 -3.82852 -2.03958 4.78748 2.74591 5.57043
    
    #  write the particle with min x to a file; it's the 10th particle
    snapmnmx p100 x min out=p100.xmin
    -4.67407 9

    # show the x,y,z,vx,vy,vz for this particle with min x
    snapprint p100.xmin
    -4.67407 1.09453 0.180351 -0.125468 -0.137605 -0.16439
.nf

.SH "SEE ALSO"
snapstat(1NEMO), snaprstat(1NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +3.0i
~/src/nbody/reduc     snapmnmx.c
.fi

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
16-Apr-91	V1.0: created          	PJT
13-may-91	V1.1: added time to mode, added more doc	PJT
8-nov-93	V1.2: using moment.h	PJT
17-aug-2022	V1.3: added out=	PJT
.fi




src/nbody/reduc/snapmnmx.c
/*
 *  SNAPMNMX: find min and/or max of snapshot bodyvars'c
 *
 *	16-Apr-91	V1.0 created      		PJT
 *      13-may-91       V1.1 added time to list of options  PJT
 *	 6-nov-93	V1.2 moment, NEMO V2.			pjt
 */

#include <stdinc.h>
#include <getparam.h>
#include <moment.h>
#include <vectmath.h>		/* otherwise NDIM undefined */
#include <filestruct.h>

#include <snapshot/snapshot.h>	
#include <snapshot/body.h>
#include <snapshot/get_snap.c>

string defv[] = {		/* DEFAULT INPUT PARAMETERS */
    "in=???\n			Input file (snapshot)",
    "var=x,y,z\n                Variables to find min/max of",
    "mode=min,max\n             Modes: {time,min,max,mean,sigma}",
    "times=all\n                Times of snapshot",
    "format=%g\n                Format to print with",
    "VERSION=1.2b\n		8-oct-08 pjt",
    NULL,
};

string usage = "find min and/or max of snapshot variables";

string cvsid = "$Id$";

#define MAXOPT    50

nemo_main()
{
    stream instr, tabstr;
    real   tsnap, ekin, etot, dr, r, rv, v, vr, vt, aux;
    real   varmin[MAXOPT], varmax[MAXOPT];
    real   var0[MAXOPT], var1[MAXOPT], var2[MAXOPT];
    Moment var[MAXOPT];
    string headline=NULL, options, times, mnmxmode;
    Body *btab = NULL, *bp, *bq;
    bool   Qmin, Qmax, Qmean, Qsig, Qtime, scanopt();
    int i, n, nbody, bits, nsep, isep, nopt, ParticlesBit;
    char fmt[20],*pfmt;
    string *burststring(), *opt;
    rproc btrtrans(), fopt[MAXOPT], faux;

    ParticlesBit = (MassBit | PhaseSpaceBit | PotentialBit | AccelerationBit |
            AuxBit | KeyBit);
    instr = stropen(getparam("in"), "r");	/* open input file */
    mnmxmode= getparam("mode");

    opt = burststring(getparam("var"),", ");
    nopt = 0;					/* count options */
    while (opt[nopt]) {				/* scan through options */
        fopt[nopt] = btrtrans(opt[nopt]);
        nopt++;
        if (nopt==MAXOPT) {
            dprintf(0,"\n\nMaximum number of var's = %d exhausted\n",MAXOPT);
            break;
        }
    }
    dprintf(0,"var: \n");
    for (i=0; i<nopt; i++)
        dprintf(0,"%s ",opt[i]);
    dprintf(0,"\n");
    dprintf(0,"mode: %s\n",mnmxmode);

    Qmin = scanopt(mnmxmode,"min");
    Qmax = scanopt(mnmxmode,"max");
    Qmean = scanopt(mnmxmode,"mean");
    Qsig = scanopt(mnmxmode,"sigma");
    Qtime = scanopt(mnmxmode,"time");
    if (!Qmin && !Qmax && !Qmean && !Qsig && !Qtime) 
        error("No mode selected");
#if 0
    pfmt = getparam("tab");
    if (pfmt!=NULL && *pfmt!=NULL) {
	dprintf(0,"Saving table in %s\n",pfmt);
        tabstr = stropen(pfmt,"w");
    } else
#endif
        tabstr = stdout;

    times = getparam("times");
    pfmt = getparam("format");
    strcpy (fmt,pfmt);
    if (strchr(fmt,' ')==NULL && strchr(fmt,',')==NULL)
        strcat (fmt," ");       /* append blank if user did not specify sep */

    get_history(instr);                 /* read history */

    for(;;) {                /* repeating until first or all times are read */
	get_history(instr);
        if (!get_tag_ok(instr, SnapShotTag))
            break;                                  /* done with work */
        get_snap(instr, &btab, &nbody, &tsnap, &bits);
        if (!streq(times,"all") && !within(tsnap,times,0.0001))
            continue;                   /* skip work on this snapshot */
        if ( (bits & ParticlesBit) == 0)
            continue;                   /* skip work, only diagnostics here */

            for (bp = btab, i=0; bp < btab+nbody; bp++, i++) {
                for (n=0; n<nopt; n++) {
                    aux = fopt[n](bp,tsnap,i);
                    if (i==0) ini_moment(&var[n],2,0);
                    accum_moment(&var[n], aux, 1.0);
                }
            }
            if (Qtime)
                fprintf(tabstr,fmt,tsnap);
            if (Qmin) {
                for (n=0; n<nopt; n++)
                    fprintf(tabstr,fmt,min_moment(&var[n]));
            }
            if (Qmax) {
                for (n=0; n<nopt; n++)
                    fprintf(tabstr,fmt,max_moment(&var[n]));
            }
            if (Qmean) {
                for (n=0; n<nopt; n++)
                    fprintf(tabstr,fmt,mean_moment(&var[n]));
            }
            if (Qsig) {
                for (n=0; n<nopt; n++)
                    fprintf(tabstr,fmt,sigma_moment(&var[n]));
            }
            fprintf(tabstr,"\n");
    
    }
    strclose(instr);
}
